File: frontend/app/models/difficulty.ts
import {z} from "zod";

/** REST API Difficulty schema. */
const ApiDifficultySchema = z.object({
    id: z.number(),
    name: z.string(),
    label: z.string(),
    max_attempts: z.number(),
    code_length: z.number(),
    minimum_number: z.number(),
    maximum_number: z.number(),
    is_duplicate_allowed: z.boolean()
});

/** Difficulty schema. */
const DifficultySchema = ApiDifficultySchema.transform((data) => ({
    id: data.id,
    name: data.name,
    label: data.label,
    maxAttempts: data.max_attempts,
    codeLength: data.code_length,
    minimumNumber: data.minimum_number,
    maximumNumber: data.maximum_number,
    isDuplicateAllowed: data.is_duplicate_allowed
}));

/** Difficulty list schema. */
export const DifficultyListSchema = z.array(DifficultySchema);

/** Difficulty type. */
export type DifficultyType = z.infer<typeof DifficultySchema>;

/** Difficulty list type. */
export type DifficultyListType = z.infer<typeof DifficultyListSchema>;

export default DifficultySchema;

File: backend/models/difficulty.py
from sqlalchemy import Column, Integer, Boolean
from sqlalchemy import Text
from sqlalchemy.orm import relationship

from models.database import Base


class Difficulty(Base):
    """ Difficulty model. """
    __tablename__ = "difficulties"

    id = Column(Integer, primary_key=True, index=True)  # Difficulty id.
    name = Column(Text, unique=True, nullable=False, index=True)  # Difficulty name is unique.
    label = Column(Text, unique=True, nullable=False, index=True)  # Difficulty label is unique.
    max_attempts = Column(Integer, nullable=False)  # Maximum attempts is required.
    code_length = Column(Integer, nullable=False)  # Code length is required.
    minimum_number = Column(Integer, nullable=False)  # Minimum number is required.
    maximum_number = Column(Integer, nullable=False)  # Maximum number is required.
    is_duplicate_allowed = Column(Boolean)  # Duplicate allowed can be true or false.

    # Relationships.
    games = relationship("Game", back_populates="difficulty")  # Difficulty has many games.
